File: LC/python/Generate Parentheses.py
class Solution:
    # @param an integer
    # @return a list of string
    def generateParenthesis(self, n):
        if n == 0:
            return []
        
        res = []
        self.helper(n,n,'',res)
        return res
        
    def helper(self,left,right,item,res):
        if right < left:
            return 
        if left==0 and right==0:
            res.append(item)
        if left > 0:
            self.helper(left-1, right, item+'(', res)
        if right > 0:
            self.helper(left, right-1, item+')',res)
        
        

File: LC/java/Generate Parentheses.java
public class Solution {
    public ArrayList<String> generateParenthesis(int n) {
        if(n<=0){
            return new ArrayList<String>();
        }
        ArrayList<String> res = new ArrayList<String>();
        dfs(res,"",n,n);
        return res;
    }
    
    public void dfs(ArrayList<String> res, String sub, int left, int right){
        if(left<0 || right<0 || left>right){
            return;
        }
        
        if(left==0 && right==0){
            res.add(sub);
            return;
        }
        
        dfs(res,sub+"(",left-1,right);
        dfs(res,sub+")",left,right-1);
        
    }
}